src/charm.py

```diff
@@ -1,103 +1,292 @@
 #!/usr/bin/env python3
-# Copyright 2023 identity-team
+# Copyright 2023 Canonical Ltd.
 # See LICENSE file for licensing details.
 #
 # Learn more at: https://juju.is/docs/sdk
 
-"""Charm the service.
+"""A Juju charm for GLAuth."""
 
-Refer to the following post for a quick-start guide that will help you
-develop a new k8s charm using the Operator Framework:
+import logging
+from pathlib import Path
+from typing import TYPE_CHECKING, Any
 
-https://discourse.charmhub.io/t/4208
-"""
+from charms.data_platform_libs.v0.data_interfaces import (
+    DatabaseCreatedEvent,
+    DatabaseEndpointsChangedEvent,
+    DatabaseRequires,
+)
+from charms.loki_k8s.v0.loki_push_api import LogProxyConsumer, PromtailDigestError
+from charms.observability_libs.v0.kubernetes_service_patch import KubernetesServicePatch
+from charms.prometheus_k8s.v0.prometheus_scrape import MetricsEndpointProvider
+from charms.traefik_k8s.v1.ingress import (
+    IngressPerAppReadyEvent,
+    IngressPerAppRequirer,
+    IngressPerAppRevokedEvent,
+)
+from jinja2 import Template
+from ops.charm import CharmBase, ConfigChangedEvent, HookEvent, PebbleReadyEvent
+from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus, ModelError, WaitingStatus
+from ops.pebble import ChangeError, Layer
 
-import logging
+from utils import normalise_url
 
-import ops
+if TYPE_CHECKING:
+    from ops.pebble import LayerDict
 
-# Log messages can be retrieved using juju debug-log
 logger = logging.getLogger(__name__)
 
-VALID_LOG_LEVELS = ["info", "debug", "warning", "error", "critical"]
 
+class GlauthK8SCharm(CharmBase):
+    """Charmed GLAuth."""
 
-class GlauthK8SCharm(ops.CharmBase):
-    """Charm the service."""
-
-    def __init__(self, *args):
+    def __init__(self, *args: Any) -> None:
         super().__init__(*args)
-        self.framework.observe(self.on.httpbin_pebble_ready, self._on_httpbin_pebble_ready)
+        self._container_name = "glauth"
+        self._container = self.unit.get_container(self._container_name)
+        self._config_dir_path = Path("/etc/config")
+        self._config_file_path = self._config_dir_path / "glauth.cfg"
+
+        self._db_name = f"{self.model.name}_{self.app.name}"
+        self._db_relation_name = "database"
+        self._db_plugin = "postgres.so"
+
+        self._prometheus_scrape_relation_name = "metrics-endpoint"
+        self._loki_push_api_relation_name = "logging"
+
+        self._glauth_service_command = f"glauth -c {self._config_file_path}"
+        self._log_dir = Path("/var/log")
+        self._log_path = self._log_dir / "glauth.log"
+
+        self._ingress_relation_name = "ingress"
+
+        self.service_patcher = KubernetesServicePatch(
+            self, [(self.app.name, int(self._ldap_port))]
```

Comment:
Shouldn't we patch the `api` port as well?

Reply:
So the API port is something we only use for prometheus metrics. I'm under the impression that the http server is something we only want to use if it's behind a proxy. With that being said I can patch that port too.